### PR TITLE
Forward-port TrackerTopology migration from 620_SLHC

### PR DESCRIPTION
**Pull request by makortel**
Migrate to TrackerTopology in tracking-related files by forward-porting commits from PRs #1026 and #1265 (in 620_SLHC branch).

No regression expected, no regression observed (tested in 740pre5 with RelValTTbar_13). 

@rovere @VinInn @boudoul @venturia  @davidlange6

### CODE REVIEW

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by civanch** [comment]
+1

File: RecoTracker/TkNavigation/test/NavigationSchoolAnalyzer.cc [hunk 1]
```diff
@@ -62,13 +58,18 @@ class NavigationSchoolAnalyzer : public edm::EDAnalyzer {
   virtual void endJob() ;
 
   std::string theNavigationSchoolName;
+  const TrackerTopology *tTopo;
+  void print(std::ostream& os,const DetLayer* dl);
+  void print(std::ostream&os, const NavigationSchool::StateType & layers);
+  void print(std::ostream&os, const NavigationSchool *nav);
```
Comment (cvuosalo) [inline RecoTracker/TkNavigation/test/NavigationSchoolAnalyzer.cc]: Please add a space for readability to std::ostream&os on lines 63 and 64, just as you did on line 62.

**Comment by cvuosalo** [comment]
+1

For #7335 a00d085a1885be6a596741b14c3638b670218317 
This PR involves small changes to tracking. 

Code changes are satisfactory.

Jenkins tests [CMSSW_7_4_X_2015-01-23-1400]
(https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-7335/2085/summary.html):
compilation, unit tests, matrix tests, static checks are all OK.

The [alternative-comparisons] (https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_7_4_X_2015-01-23-1400+7335/5689/alternative-comparisons/) show a few minor differences including this plot:
1306 0-result_strip

[validateJR] (https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_7_4_X_2015-01-23-1400+7335/5689/validateJR/) shows no differences.

Extended tests give about a dozen plots with minor, insignificant differences.

**Comment by davidlange6** [comment]
alca changes are just to remove headers, so will bypass. Complain if not ok.

**Comment by venturia** [comment]
To my understanding there should be NO difference. Can anyone follow it up with a reasonable priority?

                                                 Andrea

On Jan 26, 2015, at 5:15 , cvuosalo <notifications@github.com<mailto:notifications@github.com>>
 wrote:


+1

For #7335<https://github.com/cms-sw/cmssw/pull/7335> a00d085<https://github.com/cms-sw/cmssw/commit/a00d085a1885be6a596741b14c3638b670218317>
This PR involves small changes to tracking.

Code changes are satisfactory.

Jenkins tests CMSSW_7_4_X_2015-01-23-1400<https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-7335/2085/summary.html>:
compilation, unit tests, matrix tests, static checks are all OK.

The alternative-comparisons<https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_7_4_X_2015-01-23-1400+7335/5689/alternative-comparisons/> show a few minor differences including this plot:
[1306 0-result_strip]<https://cloud.githubusercontent.com/assets/5736159/5895219/627bde3e-a4de-11e4-9182-626dd7f307db.png>

validateJR<https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_7_4_X_2015-01-23-1400+7335/5689/validateJR/> shows no differences.

Extended tests give about a dozen plots with minor, insignificant differences.

**Comment by makortel** [comment]
Indeed there should be no difference. The DQM files from my test (RelValTTbar, 9k events) show identical histograms for this. I'll run 1306.0 myself too.

**Comment by makortel** [comment]
Curious. I ran 1306.0 in 740pre5 and pre5+PR (in the same machine):

image

and indeed there is a difference (and both of the distributions differ from https://github.com/cms-sw/cmssw/pull/7335#issuecomment-71413089, but that may be just the difference between 740pre5 and CMSSW_7_4_X_2015-01-23-1400). Investigating...

**Comment by slava77** [comment]
Hi Matti,

on difference between 740pre5 and  CMSSW_7_4_X_2015-01-23-1400:
the main one is that the beamspot changed (in CMSSW_7_4_X_2015-01-22-0200)
a full update of GTs with the new beamspot  happened only in
CMSSW_7_4_X_2015-01-23-2100.



On 1/26/15 9:29 AM, Matti Kortelainen wrote:
> Curious. I ran 1306.0 in 740pre5 and pre5+PR (in the same machine):
> 
> image
> <https://cloud.githubusercontent.com/assets/33993/5902062/a2a2aff8-a577-11e4-83bf-1e5f47514048.png>
> 
> and indeed there is a difference (and both of the distributions differ
> from #7335 (comment)
> <https://github.com/cms-sw/cmssw/pull/7335#issuecomment-71413089>, but
> that may be just the difference between 740pre5 and
> CMSSW_7_4_X_2015-01-23-1400). Investigating...
> 
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/cms-sw/cmssw/pull/7335#issuecomment-71478016>.
> 

-- 
***********************************************************************
 Vyacheslav (Slava) Krutelyov
 TAMU: Physics Dept Texas A&M MS4242, College Station, TX 77843-4242
 CERN: 42-R-027
 AIM/Skype: siava16              googleTalk: slava77@gmail.com
 (630) 291-5128  Cell (US)       +41 76 275 7116 Cell (CERN)

-----------------------------------------------------------------------

**Comment by makortel** [comment]
Thanks Slava, that is probably enough to change this distribution.

File: SimGeneral/TrackingAnalysis/plugins/TrackingTruthAccumulator.cc [hunk 2]
```diff
@@ -1058,55 +1057,6 @@ namespace // Unnamed namespace for things only used in this file
...
-
-	//---------------------------------------------------------------------------------
-	//---------------------------------------------------------------------------------
-	//----   Functions in the unnamed namespace   -------------------------------------
-	//---------------------------------------------------------------------------------
-	//---------------------------------------------------------------------------------
-
-	int LayerFromDetid( const DetId& detId )
-	{
-		if( detId.det()!=DetId::Tracker ) return 0;
```
Comment (makortel) [inline SimGeneral/TrackingAnalysis/plugins/TrackingTruthAccumulator.cc]: The difference originates from this check, and what happens is the following. In the workflow 1306.0 the output from `LayerFromDetid` and `TrackerTopology::layer()` were different for the following DetId's
```
604049961
604049962
604049963
604049964
604049965
604049966
604050097
604050098
604050099
604050100
604050101
604050102
604050209
604050210
604050211
604050212
604050213
604050214
```
For these, the `DetId::det()` is equal to `DetId::Muon`. In line above, the returned layer is 0. However, in the new code, no such check is made, and `TrackerTopology::layer()` returns happily layer 1 (and assumes that the detectors are FPIX).

The best fix would probably be to make the check above before calling `TrackerTopology::layer()`. Another possibility coming to my mind would be for `TrackerTopology` to make these checks, but that would probably add too many typically redundant(?) checks to it.

By the way, is it documented somewhere that TrackerTopology assumes that the input `DetId`s belong to `DetId::Tracker`?

Comment (davidlange6) [inline SimGeneral/TrackingAnalysis/plugins/TrackingTruthAccumulator.cc]: On Jan 27, 2015, at 8:32 AM, Matti Kortelainen <notifications@github.com>
 wrote:

> In SimGeneral/TrackingAnalysis/plugins/TrackingTruthAccumulator.cc:
> 
> > @@ -1058,55 +1057,6 @@ namespace // Unnamed namespace for things only used in this file
> >  	}
> >  
> >  
> > -
> > -	//---------------------------------------------------------------------------------
> > -	//---------------------------------------------------------------------------------
> > -	//----   Functions in the unnamed namespace   -------------------------------------
> > -	//---------------------------------------------------------------------------------
> > -	//---------------------------------------------------------------------------------
> > -
> > -	int LayerFromDetid( const DetId& detId )
> > -	{
> > -		if( detId.det()!=DetId::Tracker ) return 0;
> 
> The difference originates from this check, and what happens is the following. In the workflow 1306.0 the output from LayerFromDetid and TrackerTopology::layer() were different for the following DetId's
> 
> 604049961
> 604049962
> 604049963
> 604049964
> 604049965
> 604049966
> 604050097
> 604050098
> 604050099
> 604050100
> 604050101
> 604050102
> 604050209
> 604050210
> 604050211
> 604050212
> 604050213
> 604050214
> 
> For these, the DetId::det() is equal to DetId::Muon. In line above, the returned layer is 0. However, in the new code, no such check is made, and TrackerTopology::layer() returns happily layer 1 (and assumes that the detectors are FPIX).
> 
> The best fix would probably be to make the check above before calling TrackerTopology::layer(). Another possibility coming to my mind would be for TrackerTopology to make these checks, but that would probably add too many typically redundant(?) checks to it.
> 
> By the way, is it documented somewhere that TrackerTopology assumes that the input DetIds belong to DetId::Tracker?

Not explicitly - but it makes the same assumptions that the Si*DetId classes always have. The key being that "Tracker" is not "Tracking".

Comment (makortel) [inline SimGeneral/TrackingAnalysis/plugins/TrackingTruthAccumulator.cc]: Since there were no other comments, I submitted the proposed fix in #7445.

**Pull request merged**